service/history/hsm/tree.go

```diff
@@ -452,6 +554,47 @@ func (n *Node) Sync(incomingNode *Node) error {
 
 	// sync LastUpdateVersionedTransition last as MachineTransition can't correctly handle it.
 	n.persistence.LastUpdateVersionedTransition = incomingInternalRepr.LastUpdateVersionedTransition
+
+	if incomingInternalRepr.Deleted {
+		return n.SyncDeleted(ctx, incomingNode)
+	}
+
+	return nil
+}
+
+// SyncDeleted handles replication of deleted nodes and their operation history
+func (n *Node) SyncDeleted(ctx context.Context, incomingNode *Node) error {
```

Comment:
This is the wrong API. There's probably no need to even expose this method on a node. The replication stack can walk the tree and remove any delete paths.

Reply:
Removed